### ISSUE TITLE
Incorrect Hessian of vector 2-norm using `jax.numpy.linalg.norm`

### ISSUE THREAD

**Issue by RaphelWei** [google/jax#5262]
Merry Christmas! I was a previous `autograd` user and new to `jax`. I run into this issue when I tried to compute the Hessian of a vector 2-norm using `jax.numpy.linalg.norm`.  

Let f(x)=||x||^2, then we know the Hessian should be 2I, where I is the identity matrix.

Here is the repro:
```python
import jax.numpy as jnp
from jax import grad, hessian
from jax.numpy import linalg

f = lambda x: linalg.norm(x)**2
a = jnp.array([1.,2.,3.])
hessian(f)(a)
```
The output is not correct.
```python
DeviceArray([[ 2.0000000e+00, -2.9802322e-08,  2.9802322e-08],
             [-2.9802322e-08,  2.0000000e+00,  5.9604645e-08],
             [-4.4703484e-08, -8.9406967e-08,  2.0000000e+00]],            dtype=float32)
```

However, if I rewrite the 2-norm in the following way, then I could get the expected output:
```python
g = lambda x: jnp.sum(x**2)
hessian(g)(a)

DeviceArray([[2., 0., 0.],
             [0., 2., 0.],
             [0., 0., 2.]], dtype=float32)
```

I also tried `autograd`, which works perfectly with `autograd.numpy.linalg.norm`.
```python
import autograd.numpy as np
from autograd import grad, jacobian
from autograd.numpy import linalg

f = lambda x: linalg.norm(x)**2
a = np.array([1.,2.,3.])
hessian = jacobian(grad(f))
hessian(a)

array([[2., 0., 0.],
       [0., 2., 0.],
       [0., 0., 2.]])
```

Thank you!

**Comment by jakevdp** [comment]
Hello - your first attempt is computing the correct result up to floating point accuracy: by default JAX works in 32-bit floating point precision; a good rule-of-thumb is that in float32 precision, values of a few times 10^-8 are effectively zero.

If you want to instead compute in 64-bit precision, you can set the `enable_x64` flag; see https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#Double-(64bit)-precision

**Comment by RaphelWei** [comment]
@jakevdp Thanks!
